design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1072 @@
+  `true` indicating values should be persisted within a _record_; `false`
+  indicating values should be persisted as anonymous values.
+  
+  With a legacy default of `true` to ensure backwards compatibility of old
+  queries, and a new default of `false`, meaning new queries will default
+  to writing single value fields as anonymous values.
+  
+* `ksql.persistence.wrap.single.keys`: which provides a default for how 
+  KEY schemas with a single field should be deserialized and serialized:
+  `true` indicating keys should be persisted within a _record_; `false`
+  indicating values should be persisted as anonymous values.
+  
+  With a legacy default of `true` to ensure backwards compatibility of old
+  queries with string keys, and a new default of `false`, meaning new 
+  queries will default to writing single value fields as anonymous values.
+  
+  Note: to be added as part of the future structured keys work.
+ 
+### SQL syntax
+
+Users can override the configured defaults and control how single field
+keys and values are serialized by providing the following `WITH` clause
+properties.
+
+* `WRAP_SINGLE_VALUES`: boolean property that will override the
+  `ksql.persistence.wrap.single.values` configuration.
+
+* `WRAP_SINGLE_KEYS`: boolean property that will override the
+  `ksql.persistence.wrap.single.keys` configuration.
+  
+  Note: to be added as part of the future structured keys work.
+  
+* `WRAP_SINGLE_FIELDS`: boolean property that is a short hand for 
+  setting both `WRAP_SINGLE_KEYS` and `WRAP_SINGLE_VALUES`.
+  
+  Note: to be added as part of the future structured keys work.
+
+## Design
+
+### `CREATE STEAM` and `CREATE TABLE`
+
+C* statements will, by default, capture the values of 
+`ksql.persistence.wrap.single.keys` and `ksql.persistence.wrap.single.values`.  
+These settings will be stored as part of the created source's metadata. 
+
+Users can override these settings by providing any of 
+the `WRAP_SINGLE_XXX` familty of `WITH` clause properties.
+
+These settings control both how the data in the source's topic should be
+deserialized if it only has a single field _and_ how downstream queries
+should serialize any single fields.  
+
+Note: as the settings control both deserialization and serialization, it is
+not possible to define a new source with a single field that should be 
+_deserialized_ unwrapped, but have downstream queries _serialize_ wrapped,
+or vice-versa. However, this limitation does allow for more simple syntax
+and downstream queries can override the setting as needed. (See rejected 
+alternatives for and alternative that uses two different settings for 
+deserialization and serialization).
+
+For example:
+
+```sql
+-- Override 'ksql.persistence.wrap.single.values' to true
+-- Indicating that the value of the data in the stream is wrapped. 
+-- Downstream queries will wrap single values when serializing them.  
+CREATE STREAM FOO (ID INT) WITH (WRAP_SINGLE_VALUES=true, ...); 
```

Comment:
nit: here and in subsequent examples, i expect the "true" and "false" will be single-quoted in practice, for consistency with other WITH properties?

Reply:
it will work both quoted and unquoted. WITH properties can be any literal, i.e. string, number, boolean, etc.